pyam/core.py

```diff
@@ -1269,6 +1269,46 @@ def normalize(self, inplace=False, **kwargs):
         if not inplace:
             return ret
 
+    def offset(self, padding=0, inplace=False, **kwargs):
+        """Compute new data which is offset from a specific data point
+
+        For example, offsetting from `year=2005` will provide data
+        *relative* to `year=2005` such that the value in 2005 is 0 and
+        all other values `value[year] - value[2005]`.
+
+        Conceptually this operation performs as:
+        ```
+        df - df.filter(**kwargs) + padding
+        ```
+
+        Note: Currently only supports normalizing to a specific time.
+
+        Parameters
+        ----------
+        padding : float, optional
+            an additional offset padding
+        inplace : bool, optional
+            if :obj:`True`, do operation inplace and return None
+        kwargs
+            the column and value on which to offset (e.g., `year=2005`)
+        """
+        if len(kwargs) > 1 or self.time_col not in kwargs:
+            raise ValueError("Only time(year)-based normalization supported")
+        ret = self.copy() if not inplace else self
+        df = ret.data
+        # change all below if supporting more in the future
+        cols = self.time_col
+        value = kwargs[self.time_col]
+        x = df.set_index(IAMC_IDX)
+        x["value"] -= x[x[cols] == value]["value"]
+        x["value"] += padding
+
+        x = x.reset_index()
+        ret._data = x.set_index(self.dimensions).value
```

Comment:
Since `set_index` is an expensive factorization operation on large datasets, i would suggest to avoid it in favour of working directly on the series representation:

```suggestion
        data = ret._data
        base_value = data.loc[data.index.isin([value], level=self.time_col)].droplevel(self.time_col)
        ret._data = data - base_value + padding
```

Reply:
Caveat with both implementations: If the base value in the reference year is missing for a timeseries the whole timeseries will be nan and dropped from the iam dataframe.

Reply:
I think that the behavior "drop values if the reference is missing" is consistent with the default behavior of other methods, e.g., [`add()`](https://pyam-iamc.readthedocs.io/en/stable/api/iamdataframe.html#pyam.IamDataFrame.add).

Or we could of course add a `fillna` argument (at a later point)...

Reply:
> Or we could of course add a fillna argument (at a later point)...

That would mean the change:

```python
    def offset(..., fillna=None):
        [...]
        ret._data = data.subtract(base_value, fillna=fillna) + padding
```